src/logger.js

```diff
@@ -0,0 +1,135 @@
+const fs = require("fs");
+
+const colors = {
+  info: "\x1b[32m",
+  debug: "\x1b[35m",
+  trace: "\x1b[36m",
+  warn: "\x1b[33m",
+  error: "\x1b[31m",
+  fatal: "\x1b[1m\x1b[31m",
+  reset: "\x1b[0m"
+}
+
+/**
+ * The `Logger` class is used to keep track of events when debugging IJO.
+ * There are different types of logs which each have there own function. 
+ * 
+ * ### Log Types
+ * 
+ * - `info` - Simple information reflecting normal behaviour. Events such as the start and stop of services are included
+ * - `debug` (optional) - More granular information about the events taking place in the running enviroment
+ * - `trace` (optional) - Logs every detail that could be of use. 
+ * - `warn` - Unexpected events/uses that should not occure normally
+ * - `error` - Serious issues that can cause things to fail
+ * - `fatal` - Serious issues that cause the entire application to fail
+ */
+class Logger {
+  /**
+   * Create a new logging instance
+   * 
+   * ### Logging Level
+   * 
+   * `0` - logs `info`, `warn`, `error`, and `fatal`. Usage for production
+   * `1` - logs all in `0` and `debug`. Used for basic debugging
+   * `2` - logs all in `1` and `trace`. Used to see every detail logged for advanced debugging
+   * 
+   * @param {string} name Name of the log
+   * @param {number} log_level The level of logging accepted, see above
+   */
+  constructor(name, log_level) {
+    this.cache = "";
+    this.name = name;
+    if (log_level < 0 || log_level > 2) throw Error(`Log level of '${log_level}' not accepted; must be 0, 1, or 2`);
+  }
+
+  /**
+   * Initialize the log with a path
+   * @param {string} path Path of file to write log (should be in the ./logs/ directory)
+   */
+  initialize(path) {
+    this.path = path;
+    this.info(`Log '${this.name}' initialized at ${path}`);
+  }
+
+  /**
+   * Write to the log file
+   * @param {string} msg Message to write to file
+   */
+  async write(msg) {
+    this.cache += `[${this.timestamp()}] ${msg}\n`;
+    fs.writeFile(this.path, this.cache, "utf8", err => {
```

Comment:
Maybe appendFile might be more clear instead of writeFile. And then you only add to the cache as long as the path is undefined and in the initialize function you write what is in the cache to the file.

Reply:
Ah, so the cache is not used after initialise?

Reply:
Yeah, because then you now which file to append to.

Reply:
The idea of the system is that you can log before it is known where to log to. This helps you to keep the customization where to log to, while still being able to log at startup. If that makes sense 😄 .

Reply:
Resolved